drivers/timer/nrf_rtc_timer.c

```diff
@@ -63,7 +61,10 @@ void rtc1_nrf_isr(void *arg)
 	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
 		u32_t next = last_count + CYC_PER_TICK;
 
-		if (counter_sub(next, t) < MIN_DELAY) {
+		/* As below: we're guaranteed to get an interrupt as
+		 * long as it's set two or more cycles in the future
+		 */
+		if (counter_sub(next, t) < 3) {
```

Comment:
If CYC_PER_TICK = 1 (which is default) , the "next" may be still too close to current counter value for reliable interrupt.

IMHO we will need here something similar to the CONFIG_ZERO_LATENCY_IRQS workaround which you implemented below (next += 3).

Reply:
No, this is OK per docs.  As long as the time between the read of "t" above and this line are less than 30us (which is trivially verifiable by counting cycles).  The ZLI workaround is the failsafe for this too -- if it runs too long, we'll detect it there.

Reply:
And just to be clear: note we're masked here.  Nothing can interrupt us *except* a ZLI.  The timing under normal circumstances is 100% deterministic.

Reply:
i think that same mechanism as in `z_clock_set_timeout` should be applied here, given ZLI presence.

Reply:
The ZLI is not only problem here. Since we are in the ISR and it not tickless, our `next` equals to `current + 1` for CYC_PER_TICK = 1. After correction, the next will be `current + 2` which is still too close to the current counter state value even if we assume that execution of the ISR taken no time.

IMHO we should make the code reliable (for example by scheduling next interrupt on `now +3`, not `now + 2^24+1`)  as well as put a warning here that we cannot have such high tick rate without tickless mode.

Reply:
> After correction, the next will be t + 2 which is still too close to the current counter value

It is not, though.  The docs for both nRF51[1] and nRF52[2] are quite clear that T+2 is guaranteed to produce an interrupt.

And like I mentioned earlier, I went in and (crudely) measured this.  It's not a synchronous device, the clock domains are different, and the overlap is in fact an analog thing and comes out to about 20% of a RTC cycle on nrf52840_pca10056.  The "+2" already includes the needed padding.

[1] "_nRF51 Series Reference Manual Version 3.0_", Section 19.1.7, p. 106

[2] "_nRF52840 Product Specification v1.0_" Section 6.22.7, p. 337